Assets/Scripts/Hierarchical/HierarchicalBase.cs

```diff
@@ -117,6 +137,83 @@ private Vector3 GetMean(System.Func<IHierarchical, Vector3> selector) {
+      }
+      if (hierarchical.Target.ActiveSubHierarchicals.Count() <= capacity) {
+        return new List<IHierarchical> { hierarchical.Target };
+      }
+
+      var possibleTargets = new List<IHierarchical>();
+      foreach (var subHierarchical in hierarchical.SubHierarchicals) {
+        possibleTargets.AddRange(FindPossibleHierarchicalTargets(subHierarchical));
+      }
+      return possibleTargets;
+    }
+    List<IHierarchical> possibleTargets = FindPossibleHierarchicalTargets(this);
+    if (possibleTargets.Count == 0) {
+      return null;
+    }
+
+    // Use a maximum speed assignment to select the target resulting in the maximum intercept speed.
+    IAssignment targetAssignment =
+        new MaxSpeedAssignment(Assignment.Assignment_EvenAssignment_Assign);
+    List<AssignmentItem> assignments =
+        targetAssignment.Assign(new List<IHierarchical> { hierarchical }, possibleTargets);
+    if (assignments.Count != 1) {
+      return null;
+    }
+    return assignments[0].Second;
+  }
+
+  private IHierarchical FindBestLeafHierarchicalTarget(IHierarchical hierarchical, int capacity) {
+    List<IHierarchical> FindLeafHierarchicalTargets(IHierarchical hierarchical) {
+      // Traverse the agent hierarchy to find only the leaf hierarchical objects.
+      if (hierarchical.SubHierarchicals.Count > 0) {
+        var leafHierarchicalTargets = new List<IHierarchical>();
+        foreach (var subHierarchical in hierarchical.ActiveSubHierarchicals) {
+          leafHierarchicalTargets.AddRange(FindLeafHierarchicalTargets(subHierarchical));
+        }
+        return leafHierarchicalTargets;
+      }
+
+      if (hierarchical.Target == null) {
+        return new List<IHierarchical>();
+      }
+      return new List<IHierarchical> { hierarchical.Target };
+    }
+    List<IHierarchical> leafHierarchicalTargets = FindLeafHierarchicalTargets(this);
+    if (leafHierarchicalTargets.Count == 0) {
+      return null;
+    }
+
+    // Use a maximum speed assignment to select the target resulting in the maximum intercept speed.
+    IAssignment targetAssignment =
+        new MaxSpeedAssignment(Assignment.Assignment_EvenAssignment_Assign);
+    List<AssignmentItem> assignments =
+        targetAssignment.Assign(new List<IHierarchical> { hierarchical }, leafHierarchicalTargets);
+    if (assignments.Count != 1) {
+      return null;
+    }
+
+    // Remove as many target sub-hierarchical objects until the interceptor capacity.
+    var targetSubHierarchicals = assignments[0].Second.ActiveSubHierarchicals;
+    var filteredSubHierarchicals =
+        targetSubHierarchicals
+            .OrderBy(subHierarchical =>
+                         Vector3.Distance(subHierarchical.Position, assignments[0].Second.Position))
+            .Take(capacity);
+    var targetHierarchical = new HierarchicalBase();
+    foreach (var subHierarchical in filteredSubHierarchicals) {
+      targetHierarchical.AddSubHierarchical(subHierarchical);
+    }
```

Comment:
is this guy an orphan? its never added to the overall hierarchy structure, doesnt have a `Target`, but the filtered subhiers remain in their original parent's list so there is ownership ambiguity? No pursuer tracking is maintained for this "synthetic" hier object.

Wonder if this orphan should be integrated into the hierarchy properly?

Reply:
There are two hierarchies: one for the interceptors (the agent hierarchy) and one to manage the threats and their clustering.  This `targetHierarchical` is for a threats, so it's assigned as the target of an interceptor.